cmd/kubeadm/app/cmd/certs_test.go

```diff
@@ -482,3 +494,229 @@ kubernetesVersion: %s`,
 		})
 	}
 }
+
+func TestRunCmdCertsExpiration(t *testing.T) {
+	kdir := testutil.SetupTempDir(t)
+	defer func() {
+		if err := os.RemoveAll(kdir); err != nil {
+			t.Fatalf("Failed to teardown: %s", err)
+		}
+		clientSetFromFile = kubeconfigutil.ClientSetFromFile
+	}()
+
+	cfg := testutil.GetDefaultInternalConfig(t)
+	cfg.CertificatesDir = kdir
+
+	// Generate all the CA
+	caCerts := map[string]*x509.Certificate{}
+	caKeys := map[string]crypto.Signer{}
+	for _, ca := range []*certsphase.KubeadmCert{
+		certsphase.KubeadmCertRootCA(),
+		certsphase.KubeadmCertFrontProxyCA(),
+		certsphase.KubeadmCertEtcdCA(),
+	} {
+		caCert, caKey, err := ca.CreateAsCA(cfg)
+		if err != nil {
+			t.Fatalf("couldn't write out CA %s: %v", ca.Name, err)
+		}
+		caCerts[ca.Name] = caCert
+		caKeys[ca.Name] = caKey
+	}
+
+	// Generate all the signed certificates
+	kubeadmCerts := []*certsphase.KubeadmCert{
+		certsphase.KubeadmCertAPIServer(),
+		certsphase.KubeadmCertKubeletClient(),
+		certsphase.KubeadmCertFrontProxyClient(),
+		certsphase.KubeadmCertEtcdAPIClient(),
+		certsphase.KubeadmCertEtcdServer(),
+		certsphase.KubeadmCertEtcdPeer(),
+		certsphase.KubeadmCertEtcdHealthcheck(),
+	}
+	for _, cert := range kubeadmCerts {
+		caCert := caCerts[cert.CAName]
+		caKey := caKeys[cert.CAName]
+		if err := cert.CreateFromCA(cfg, caCert, caKey); err != nil {
+			t.Fatalf("couldn't write certificate %s: %v", cert.Name, err)
+		}
+	}
+
+	// Generate all the kubeconfig files with embedded certs
+	kubeConfigs := []string{
+		kubeadmconstants.AdminKubeConfigFileName,
+		kubeadmconstants.SuperAdminKubeConfigFileName,
+		kubeadmconstants.SchedulerKubeConfigFileName,
+		kubeadmconstants.ControllerManagerKubeConfigFileName,
+	}
+	for _, kubeConfig := range kubeConfigs {
+		if err := kubeconfigphase.CreateKubeConfigFile(kubeConfig, kdir, cfg); err != nil {
+			t.Fatalf("couldn't create kubeconfig %q: %v", kubeConfig, err)
+		}
+	}
+
+	// A minimal kubeadm config with just enough values to avoid triggering
+	// auto-detection of config values at runtime.
+	var kubeadmConfig = fmt.Sprintf(`
+apiVersion: %[1]s
+kind: ClusterConfiguration
+certificatesDir: %s
+kubernetesVersion: %s`,
+		kubeadmapiv1.SchemeGroupVersion.String(),
+		cfg.CertificatesDir,
+		kubeadmconstants.MinimumControlPlaneVersion.String())
+
+	// Write the minimal kubeadm config to a file
+	customConfigPath := kdir + "/kubeadm.conf"
+	f, err := os.Create(customConfigPath)
+	require.NoError(t, err)
+	_, err = f.Write([]byte(kubeadmConfig))
+	require.NoError(t, err)
+
+	// fakeClientSetFromFile returns a fake clientset with kubeadm config map
+	var fakeClientSetFromFile = func(_ string) (kubernetes.Interface, error) {
+		objects := []runtime.Object{
+			testClusterConfigMap(string(kubeadmConfig), false),
+		}
+		return fakeclientset.NewSimpleClientset(objects...), nil
```

Comment:
instead of doing this

we can use reactors
see:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/apiclient/idempotency_test.go#L158-L160

- create a fake client
- add "get" / "configmaps" reactor that returns a CM.